src/main/kotlin/org/rust/cargo/project/workspace/impl/CargoProjectWorkspaceImpl.kt

```diff
@@ -138,54 +154,102 @@ class CargoProjectWorkspaceImpl(private val module: Module) : CargoProjectWorksp
                     }
                 })
 
-                Result.Ok(description)
+                UpdateResult.Ok(description)
             } catch (e: ExecutionException) {
-                Result.Err(e)
+                UpdateResult.Err(e)
             }
         }
 
         override fun onSuccess() {
-            val result = requireNotNull(result)
+            val r = requireNotNull(result)
 
-            when (result) {
-                is Result.Err -> {
-                    showBalloon("Project '${module.project.name}' update failed: ${result.error.message}", MessageType.ERROR)
-                    LOG.info("Project '${module.project.name}' update failed", result.error)
+            lock (lock) {
+                cached = when (r) {
+                    is UpdateResult.Ok -> r.projectDescription
+                    else               -> null
                 }
-                is Result.Ok  ->
+
+                notifyCargoProjectUpdate(r)
+            }
+
+            when (r) {
+                is UpdateResult.Ok  -> LOG.info("Project '${module.project.name}' successfully updated")
+                is UpdateResult.Err -> LOG.info("Project '${module.project.name}' update failed", r.error)
+            }
+        }
+    }
+
+    /**
+     * Project-update task's notifier
+     */
+    inner class Notifier : CargoProjectWorkspaceListener {
```

Comment:
This one is tricky actually. I think that showing balloons after automatic updates is a bad idea. Positive result does not give any information, and negative result can legitimately happen if, say, the user saved `Cargo.toml` in the middle of editing. 

On the other hand, showing feedback after explicitly requested update is the right thing to do.

So I think a `Listener` not a right choice here, because, first of all, you can just use a direct method call and, second, the *caller* of the `requestUpdate` must decide if the balloon is necessary.